e2e/playwright/conftest.py

```diff
@@ -0,0 +1,419 @@
+
+import contextlib
+import os
+import re
+import shlex
+import shutil
+import socket
+import subprocess
+import sys
+import time
+from io import BytesIO
+from pathlib import Path
+from tempfile import TemporaryFile
+from types import ModuleType
+from typing import Any, Generator, List, Literal, Protocol
+
+import pytest
+import requests
+from PIL import Image
+from playwright.sync_api import ElementHandle, Locator, Page
+from pytest import FixtureRequest
+
+
+class AsyncSubprocess:
+    """A context manager. Wraps subprocess.Popen to capture output safely."""
+
+    def __init__(self, args, cwd=None, env=None):
+        self.args = args
+        self.cwd = cwd
+        self.env = env
+        self._proc = None
+        self._stdout_file = None
+
+    def terminate(self):
+        """Terminate the process and return its stdout/stderr in a string."""
+        # Terminate the process
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc.wait()
+            self._proc = None
+
+        # Read the stdout file and close it
+        stdout = None
+        if self._stdout_file is not None:
+            self._stdout_file.seek(0)
+            stdout = self._stdout_file.read()
+            self._stdout_file.close()
+            self._stdout_file = None
+
+        return stdout
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def start(self):
+        # Start the process and capture its stdout/stderr output to a temp
+        # file. We do this instead of using subprocess.PIPE (which causes the
+        # Popen object to capture the output to its own internal buffer),
+        # because large amounts of output can cause it to deadlock.
+        self._stdout_file = TemporaryFile("w+")
+        print(f"Running: {shlex.join(self.args)}")
+        self._proc = subprocess.Popen(
+            self.args,
+            cwd=self.cwd,
+            stdout=self._stdout_file,
+            stderr=subprocess.STDOUT,
+            text=True,
+            env={**os.environ.copy(), **self.env} if self.env else None,
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc = None
+        if self._stdout_file is not None:
+            self._stdout_file.close()
+            self._stdout_file = None
+
+
+def resolve_test_to_script(test_module: ModuleType) -> str:
+    """Resolve the test module to the corresponding test script filename."""
+    module_name = test_module.__name__
+    return module_name.replace("_test", ".py")
+
+
+def find_available_port(host: str = "localhost") -> int:
+    """Find an available port on the given host."""
+    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind((host, 0))  # 0 means that the OS chooses a random port
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return int(s.getsockname()[1])
```

Comment:
nit: Can we add a comment what `[1]` getting? From what I researched, it seems to be the port but should maybe add a comment:
`# get port number`

Reply:
I added a comment :)